cdc_iterator.go

```diff
@@ -0,0 +1,303 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-mysql/common"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/go-mysql-org/go-mysql/canal"
+	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/go-mysql-org/go-mysql/schema"
+	mysqldriver "github.com/go-sql-driver/mysql"
+)
+
+type cdcIterator struct {
+	canal       *common.Canal
+	rowsEventsC chan rowEvent
+
+	canalRunErrC chan error
+	canalDoneC   chan struct{}
+
+	config cdcIteratorConfig
+}
+
+type cdcIteratorConfig struct {
+	tables      []string
+	mysqlConfig *mysqldriver.Config
+	position    sdk.Position
+	TableKeys   common.TableKeys
+}
+
+func newCdcIterator(ctx context.Context, config cdcIteratorConfig) (common.Iterator, error) {
+	c, err := common.NewCanal(config.mysqlConfig, config.tables)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create canal: %w", err)
+	}
+
+	sdk.Logger(ctx).Info().Msg("created canal")
+
+	pos, err := c.GetMasterPos()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get master pos when creating cdc iterator: %w", err)
+	}
```

Comment:
Can we send this position to `getStartPosition` so we don't do the same round trip to the DB twice? Also it will prevent a possible mismatch between this position and the start position.

Reply:
Ooh good catch, thanks a lot @lovromazgon